nuagekubemon/client/nuageetcdclient.go

```diff
@@ -0,0 +1,630 @@
+				if err != nil {
+					glog.Errorf("get on %s failed: %v", SUBNET_METADATA_TREE+ns+"/"+subnet, err)
+					return err
+				}
+				return nil
+			})
+		if err != nil {
+			return err
+		}
+		if len(getResp.Kvs) == 0 {
+			return nil
+		}
+		nuageEtcdRetry(
+			func() error {
+				_, err = nuageetcd.client.Delete(context.Background(), SUBNET_METADATA_TREE+ns+"/"+subnet, clientv3.WithPrefix())
+				if err != nil {
+					glog.Errorf("delete on %s failed: %v", SUBNET_METADATA_TREE+ns+"/"+subnet, err)
+					return err
+				}
+				return nil
+			})
+		if err != nil {
+			return err
+		}
+		return nil
+	}, SUBNET_METADATA_TREE+"subnet-lock"); err != nil {
+		glog.Errorf("deleting last subnet %s failed: %v", SUBNET_METADATA_TREE+ns+"/"+subnet, err)
+		return err
+	}
+	return nil
+}
+
+//AllocateSubnet marks an entry if the subnet is not already used
+func (nuageetcd *NuageEtcdClient) AllocateSubnetCIDR(subnet *api.EtcdSubnetMetadata) (string, error) {
+	var err error
+	var getResp *clientv3.GetResponse
+	allocatedSubnet := ""
+	networkCIDR := strings.Replace(subnet.CIDR, "/", "-", -1)
+	err = nuageetcd.nuageMutex(func() error {
+		nuageEtcdRetry(
+			func() error {
+				getResp, err = nuageetcd.client.Get(context.Background(), POOL_CIDR_TREE, clientv3.WithPrefix())
+				if err != nil {
+					glog.Errorf("fetching all network cidrs from etcd failed: %v", err)
+					return err
+				}
+				return nil
+			})
+		if err != nil {
+			return err
+		}
+		for _, kv := range getResp.Kvs {
+			if networkCIDR == path.Base(string(kv.Key)) {
+				allocatedSubnet = string(kv.Value)
+				return nil
+			}
+		}
+		nuageEtcdRetry(
+			func() error {
+				_, err = nuageetcd.client.Put(context.Background(), POOL_CIDR_TREE+networkCIDR, subnet.Name)
+				if err != nil {
+					glog.Errorf("allocating new subnet failed with error: %v", err)
+					return err
+				}
+				return nil
+			})
+		if err != nil {
+			return err
+		}
+		return nil
+	}, POOL_CIDR_TREE+"cidr-lock")
+	return allocatedSubnet, err
+}
+
+//FreeSubnet delete the subnet entry
+func (nuageetcd *NuageEtcdClient) FreeSubnetCIDR(subnet *api.EtcdSubnetMetadata) error {
+	var err error
+	networkCIDR := strings.Replace(subnet.CIDR, "/", "-", -1)
+	return nuageetcd.nuageMutex(func() error {
+		nuageEtcdRetry(
+			func() error {
+				_, err = nuageetcd.client.Delete(context.Background(), POOL_CIDR_TREE+networkCIDR)
+				if err != nil {
+					glog.Errorf("deleting subnet failed with error: %v", err)
+					return err
+				}
+				return nil
+			})
+		if err != nil {
+			return err
+		}
+		return nil
+	}, POOL_CIDR_TREE+"cidr-lock")
+}
+
+//UpdateSubnetIDKey create a new key that maps to vsd network id
+func (nuageetcd *NuageEtcdClient) UpdateSubnetInfo(subnetInfo *api.EtcdSubnetMetadata) error {
+	s := etcdSubnetValue{ACTIVEIP: 0, VSDID: subnetInfo.ID, CIDR: subnetInfo.CIDR}
+	b, err := json.Marshal(s)
+	if err != nil {
+		glog.Errorf("marshal struct(%v) to string failed: %v", s, err)
+		return err
+	}
+	nuageEtcdRetry(
```

Comment:
acquiring a lock is not really adding any advantage. no other thread will update the same key. It is the same thread that creates the subnet on VSD also updates the subnet info. So i guess no explicit locking is required.